packages/demos/stories/DragDrop.tsx

```diff
@@ -0,0 +1,135 @@
+/**
+ * © 2020 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayList from '@clayui/list';
+import {XYCoord} from 'dnd-core';
+import update from 'immutability-helper';
+import React, {useState, useCallback, useRef} from 'react';
+import {useDrag, useDrop, DropTargetMonitor} from 'react-dnd';
+
+export interface IDraggableListItemProps {
+	/**
+	 * The text displayed in the List Item
+	 */
+	text: string;
+	/**
+	 * Index of the List Item in order to keep track of it's positioning within the array
+	 */
+	index: number;
+	/**
+	 * Function that handles the moving of List Items
+	 */
+	moveListItem: (dragIndex: number, hoverIndex: number) => void;
+}
+
+interface IDragItem {
+	index: number;
+	type: string;
+}
+
+const DraggableListItem: React.FC<IDraggableListItemProps> = ({
+	index,
+	moveListItem,
+	text,
+}) => {
+	const ref = useRef<HTMLDivElement>(null);
+
+	const [, drop] = useDrop({
+		accept: 'listItem',
+		hover(item: IDragItem, monitor: DropTargetMonitor) {
+			if (!ref.current) {
+				return;
+			}
+
+			const dragIndex = item.index;
+
+			const hoverIndex = index;
+
+			if (dragIndex === hoverIndex) {
+				return;
+			}
+
+			const hoverBoundingRect = ref.current!.getBoundingClientRect();
+
+			const verticalMiddle =
+				(hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;
+
+			const mousePosition = monitor.getClientOffset();
+
+			const pixelsToTop =
+				(mousePosition as XYCoord).y - hoverBoundingRect.top;
+
+			if (dragIndex < hoverIndex && pixelsToTop < verticalMiddle) {
```

Comment:
it might be helpful to set `dragIndex < hoverIndex && pixelsToTop < verticalMiddle` as a named variable so that reading through the code is a little clearer as to why you would do an early return.